nupic/research/frameworks/vernon/mixins/noise_robustness_test.py

```diff
@@ -0,0 +1,146 @@
+# with Numenta, Inc., for a separate license for this software code, the
+# following terms and conditions apply:
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Affero Public License version 3 as
+# published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+# See the GNU Affero Public License for more details.
+#
+# You should have received a copy of the GNU Affero Public License
+# along with this program.  If not, see http://www.gnu.org/licenses.
+#
+# http://numenta.org/licenses/
+# ----------------------------------------------------------------------
+
+import sys
+from functools import partial
+
+import torch
+import torch.nn.functional as F
+from torchvision import transforms
+
+from nupic.research.frameworks.pytorch.model_utils import evaluate_model
+
+
+class NoiseRobustnessTest:
+    """
+    Replaces the standard evaluate model function with a loop over the same function
+    that conducts a noise robustness test on each iteration. In the test, validation
+    accuracy is reported for samples with varying percentages of Gaussian
+    noise. The output of this mixin is a curve with noise percentage on the x axis
+    and validation accuracy on the y axis.
+    """
+
+    def setup_experiment(self, config):
+        """
+        :param config:
+            - noise_levels: Optional, a list of floats between 0 and 1 which specify
+            how many units of the input image will receive Gaussian noise. Note that
+            the zero noise case is included by default
+        """
+        super().setup_experiment(config)
+        noise_levels = config.get("noise_levels", torch.arange(0.1, 1, 0.1))
+        assert all(
+            0 < noise <= 1 for noise in noise_levels
+        ), "Noise levels must be between (0, 1]"
+        noise_mean = config.get("noise_mean", 0)
+        noise_std = config.get("noise_std", 1.0)
+        default_evaluate_model_func = config.get("evaluate_model_func", evaluate_model)
+        self.evaluate_model = partial(
+            evaluate_model_with_noise,
+            evaluate_model_func=default_evaluate_model_func,
+            noise_levels=noise_levels,
+            noise_mean=noise_mean,
+            noise_std=noise_std,
+        )
+
+    @classmethod
+    def get_execution_order(cls):
+        eo = super().get_execution_order()
+        eo["setup_experiment"].append("NoiseRobustnessTest: Initialize")
+        return eo
+
+
+def evaluate_model_with_noise(
+    model,
+    loader,
+    device,
+    batches_in_epoch=sys.maxsize,
+    criterion=F.nll_loss,
+    complexity_loss_fn=None,
+    active_classes=None,
+    progress=None,
+    post_batch_callback=None,
+    transform_to_device_fn=None,
+    evaluate_model_func=None,
+    noise_levels=None,
```

Comment:
You should make the default value `noise_levels=[]`, since `None` is not iterable and will give an error when the loop executes.

Reply:
Using an empty list as a default parameter seems to go against the default style guidelines:
/Users/acuozzo/nta/nupic.research/nupic/research/frameworks/vernon/mixins/noise_robustness_test.py:89:18: B006 Do not use mutable data structures for argument defaults.  They are created during function definition time. All calls to the function reuse this one instance of that data structure, persisting changes between them.

Instead, here is the solution that I have opted for which is consistent with style guidelines:
`if noise_levels is None: 
noise_levels = [] `

Also, note that the config defaults to the range(0, 1) in increments of 0.1 in the case that noise_levels is not specified.